dmscripts/export_dos_suppliers.py

```diff
@@ -0,0 +1,275 @@
+
+            return dict(record,
+                        declaration=supplier_framework['declaration'],
+                        onFramework=supplier_framework['onFramework'])
+        except HTTPError as e:
+            if e.status_code != 404:
+                raise
+
+    return inner
+
+
+def add_draft_counts(client, framework_slug):
+    def inner(record):
+        counts = {status: {lot: 0 for lot in LOTS} for status in DRAFT_STATUSES}
+
+        for draft in client.find_draft_services_iter(record['supplier']['id'], framework=framework_slug):
+            if draft['status'] == 'submitted':
+                counts['completed'][draft['lot']] += 1
+            elif draft['status'] == 'failed':
+                counts['failed'][draft['lot']] += 1
+            else:
+                counts['draft'][draft['lot']] += 1
+
+        return dict(record, counts=counts)
+
+    return inner
+
+
+def get_declaration_questions(declaration_content, record):
+    for section in declaration_content:
+        for question in section.questions:
+            yield question, record['declaration'].get(question.id)
+
+
+def is_incorrect_mandatory_question(question, answer):
+    if question.number in CORRECT_DECLARATION_RESPONSE_MUST_BE_TRUE and answer is not True:
+        return True
+    if question.number in CORRECT_DECLARATION_RESPONSE_MUST_BE_FALSE and answer is not False:
+        return True
+    if question.number in CORRECT_DECLARATION_RESPONSES:
+        if answer not in CORRECT_DECLARATION_RESPONSES[question.number]:
+            return True
+
+    return False
+
+
+def add_failed_questions(declaration_content):
+    def inner(record):
+        if record['declaration'].get('status') != 'complete':
+            return dict(record,
+                        failed_mandatory=['INCOMPLETE'],
+                        discretionary=[])
+
+        declaration_questions = list(get_declaration_questions(declaration_content, record))
+
+        failed_mandatory = [
+            "Q{}".format(question.number)
+            for question, answer in declaration_questions
+            if is_incorrect_mandatory_question(question, answer)
+        ]
+
+        discretionary = [
+            ("Q{}".format(question.number), answer)
+            for question, answer in declaration_questions
+            if question.number in CORRECT_DECLARATION_RESPONSE_SHOULD_BE_FALSE + MITIGATING_FACTORS
+        ]
+
+        return dict(record,
+                    failed_mandatory=failed_mandatory,
+                    discretionary=discretionary)
+
+    return inner
+
+
+def find_suppliers_with_details(client, content_loader, framework_slug):
+    pool = ThreadPool(30)
+
+    content_loader.load_manifest(framework_slug, 'declaration', 'declaration')
+    declaration_content = content_loader.get_manifest(framework_slug, 'declaration')
+
+    records = find_suppliers(client, framework_slug)
+    records = filter(None, records)
```

Comment:
Not sure why we need this filter? I don't think the API will return any empty supplier IDs?